department-of-veterans-affairs/caseflow
# BillyS/APPEALS-7680: Create Return appeal to VHA Caregiver Support Program task action

Resolves [APPEALS-7680](https://vajira.max.gov/browse/APPEALS-7680)

### Description
The "Create Document ready for Board Intake Review task action modal was added to allow appeal to be sent to board intake for docketing. As a CSP user, I need to option to send appeals to board intake when they are ready to be docketed.

### Acceptance Criteria
- [x] The following occurs when the "Return appeal to VHA Caregiver Support Program" task action is selected
   - [x] A modal appears to confirm intent
   - [x] Modal Header: Return Appeal to VHA Caregiver Support Program
   - [x] Modal informational text: If you are unable to docket this appeal due to insufficient documentation, you may return this 
          to VHA Caregiver Support Program. Provide instructions and context for this action:
   - [x] BVA Intake user is required to provide a return reason
     - [x] Field type: required free text field  
 - [x] Modal contains the following selectable options
    - [x] Cancel
      - [x] User is returned to case details screen and selection is not retained
    - [x] Return
      - [x] Select to execute task action
      - [x] Option disabled until required field(s) populated
- [x] The following occurs when the "Return appeal to VHA Caregiver Support Program" task action is executed 
   - [x] The predocket task is placed on hold until appeal is returned to Board Intake
   - [x] A VHA CSP Review Documentation task is created and is active
   - [x] The appeal appears on the Unassigned tab of the VHA CSP predocket queue
   - [x] The appeal appears on the Pending tab of BVA Intake organization queue
   - [x] BVA Intake user is returned to the BVA Intake organization queue 
     - [x] A success message appears confirming the case has been returned to Caregiver Support Program
   - [x] Return appeal to VHA Caregiver Support Program is no longer an available task action selection for the case

### Testing Plan

```ruby
1. Open the Rails console (`make c` in VSCode Terminal)

2. Create appeal with a Pre-Docket Task
     appeal = FactoryBot.create(:appeal, :with_pre_docket_task)

3. Check task tree to make sure it created a root task and Pre-docket Task
    appeal.treee

    should display:
```
   ![image](https://user-images.githubusercontent.com/98414141/182461698-60c4200d-7d46-4037-bf40-80c126193936.png)
```ruby 
    NOTE the ID on the PreDocketTask for future use.  The example displayed is 7134
 
4. Define the task by the Task ID
    task=Task.find(7134)

5. Create the Vha Document Search Task
      vha_document_search_task=VhaDocumentSearchTask.create!( parent: task, appeal: appeal, assigned_at: Time.zone.now, assigned_to: VhaCaregiverSupport.singleton )

6. Check task tree to make sure it created a Vha Docuement Search Task
      appeal.treee

     should display:
```
   ![image](https://user-images.githubusercontent.com/98414141/182461939-10a4fdea-c6d9-4882-9a01-f0e23b9ef062.png)

      This time, please NOTE the ID on the VhaDocketSearchTask for future use.  The example displayed is 
       7135. In addition, note that the status on the VhaDocumentSearchTask is "assigned"
```ruby
7.  I find this next step of setting the veteran_file_number helpful for easy identification.  
       appeal.update( veteran_file_number: "500000012" )

8.  Change the task status to "Completed" so that it will appear on the "In Progress" tab of the Intake Queue.  (Note: When all 
     the feature work is complete and all task actions are available, we would simply go into the Caregiver Queue and select the 
     task action "Documents Ready for Board Intake Review"
     
      Task.find(7135).completed!
```
9. Open Caseflow App

10. Log in as Intake User
![image](https://user-images.githubusercontent.com/98414141/185998025-087f720a-6ce8-4726-8d3c-e371e814f805.png)

11. Select Your Queue

![image](https://user-images.githubusercontent.com/98414141/182458385-e6e710b3-f3b8-41ed-b8f6-e090de6e481d.png)

12. Switch to the BVA Intake queue:
![image](https://user-images.githubusercontent.com/98414141/185998744-f99e97fb-c8f0-4af4-a21f-4ab9448de98f.png)

13. You will end up in the BVA Intake Queue.  Please note that the absence of the appeal you created on the Pending tab:
 
![image](https://user-images.githubusercontent.com/98414141/186002245-eff9e4db-a6cf-4c86-b1b0-11b63d3b76d8.png)

14. Switch to Ready for Review tab and note the presence of appeal that was created:

![image](https://user-images.githubusercontent.com/98414141/186002814-95ff494e-e698-48cd-a037-c13fd83fa1cd.png)

15. Click on the appeal and then click on the dropdown and verify the task action "Return Appeal to VHA Caregiver Support 
      Program is available:

![image](https://user-images.githubusercontent.com/98414141/186003232-77829ad0-53da-4ab8-a2e2-8e564163a305.png)

16. Select task action "Return Appeal to VHA Caregiver Support Program" and verify the following Modal appears with the 
     correct language, a cancel button and a disabled Return button:

![image](https://user-images.githubusercontent.com/98414141/186004302-75a86e91-b021-4577-8085-38c1b88ac018.png)

17. Click Cancel Button and verify you are returned to the Case details page

18. Repeat step 16

19.  This time, type text into the text box and verify that the Return button is now enabled 

![image](https://user-images.githubusercontent.com/98414141/186003872-a09d6ec6-431d-42af-8c5d-3d9aca2c376d.png)

20. Click "Return".  Verify that you are returned to the "Pending" tab of the VBA Intake Queue with the appeal listed and the 
      following confirmation message:

![image](https://user-images.githubusercontent.com/98414141/186005342-b29659df-3dec-4f8b-b97a-0df048735662.png)

21. Click on the appeal and note that the only task action available now is the "Docket Appeal" Task action

![image](https://user-images.githubusercontent.com/98414141/186005521-25b07919-fd74-48bf-b1d5-8e4a3f664feb.png)

22. Switch user to one of the Caregiver users
![image](https://user-images.githubusercontent.com/98414141/186007692-a7ca39dc-360a-469d-90dd-d4dcab83253f.png)
![image](https://user-images.githubusercontent.com/98414141/182457704-7b5dbe54-8745-489b-bb46-c5964ad0b2a5.png)

23. Select Your Queue

![image](https://user-images.githubusercontent.com/98414141/182458385-e6e710b3-f3b8-41ed-b8f6-e090de6e481d.png)

24. Switch to the CSP organization's queue:
<img width="852" alt="switch-to-csp-queue" src="https://user-images.githubusercontent.com/99351305/181598516-51ff9792-ba56-4bcd-b9fd-e8aa3776954e.PNG">

25. You will end up in the Caregiver Queue with all 3 tabs visible.  Verify you have been routed to the Unassigned tab and that 
      the appeal is listed:

![image](https://user-images.githubusercontent.com/98414141/186009435-20be88c1-f8bf-49cd-a9c9-ccb3e19dcb41.png)
